Tchap/Managers/UniversalLink/UniversalLinkService.swift

```diff
@@ -0,0 +1,184 @@
+        }
+        return false
+    }
+    
+    func handleFragment(_ fragment: String, completion: @escaping (MXResponse<UniversalLinkServiceParsingResult>) -> Void) -> Bool {
+        NSLog("[UniversalLinkService] handleFragment: \(fragment)")
+        
+        // Extract required parameters from the link
+        let params = parseFragment(fragment)
+        if params.pathParams.isEmpty {
+            NSLog("[UniversalLinkService] handleFragment: Error: No path parameters")
+            return false
+        }
+        
+        var isSupported: Bool = false
+        
+        // Check whether this is a registration links.
+        if params.pathParams[0] == Constants.registerPathParam, let registerParams = params.queryParams {
+            NSLog("[UniversalLinkService] handleFragment: link with registration parameters")
+            completion(MXResponse.success(.registrationLink(params: registerParams)))
+            isSupported = true
+        } else if params.pathParams[0] == Constants.roomPermalinkPathParam, params.pathParams.count >= 2 {
+            NSLog("[UniversalLinkService] handleFragment: link with room parameters")
+            // The link is the form of "/room/[roomIdOrAlias]" or "/room/[roomIdOrAlias]/[eventId]"
+            let roomIdOrAlias = params.pathParams[1]
+            var eventId: String?
+            
+            // Is it a link to an event of a room?
+            if params.pathParams.count == 3 {
+                eventId = params.pathParams[2]
+            }
+            
+            completion(MXResponse.success(.roomLink(roomIdOrAlias, eventID: eventId)))
+            isSupported = true
+        } else if params.pathParams[0].hasPrefix("#") || params.pathParams[0].hasPrefix("!") {
+            NSLog("[UniversalLinkService] handleFragment: link with room parameters")
+            // The link is the form of "/[roomIdOrAlias]" or "/[roomIdOrAlias]/[eventId]"
+            // Such links come from matrix.to permalinks
+            let roomIdOrAlias = params.pathParams[0]
+            var eventId: String?
+            
+            // Is it a link to an event of a room?
+            if params.pathParams.count == 2 {
+                eventId = params.pathParams[1]
+            }
+            completion(MXResponse.success(.roomLink(roomIdOrAlias, eventID: eventId)))
+            isSupported = true
+        } else {
+            NSLog("[UniversalLinkService] handleFragment: Do not know what to do with the link arguments: %@", params.pathParams)
+        }
+        
+        return isSupported
+    }
+    
+    private func parseFragment(_ fragment: String) -> UniversalLinkParameters {
+        var pathParams = [String]()
+        var queryParams: [String: String]?
+        
+        let fragmentParts = fragment.split(separator: "?")
+        
+        // Extract path params by removing empty ones, and removing percent encoding
+        let pathSubstrings = fragmentParts[0].split(separator: "/").filter { !$0.isEmpty }
```

Comment:
This could crash `fragmentParts` could be empty. Ensure `fragmentParts` is not empty or use `fragmentParts.first`

Reply:
I consider `fragmentParts` could not be empty. It should contain at least the provided `fragment`. Don't you think so?
I may use `fragmentParts.first` if you prefer

Reply:
I you give an empty string to `func parseFragment(_ fragment: String)` `fragmentParts` will be empty and `fragmentParts[0]` will crash. You can check there is one element by using:
`if let firstFragment = fragmentParts.first` for ex or `firstFragment.isEmpty == false`

Reply:
ok